variable metadata-based independent axis range

moontrip opened the pull request:
This is for visualization plots, especially histogram and scatter/line plots: continuous data. Now, they use variable's metadata-based independent axis range: either displayRangeMin/Man or rangeMin/Max defined at each variable. That is, regardless of filtering, deselection by legend item click, etc., the independent axis range remains the same. Both numeric and date type are handled. Note that for scatter plot, additional 5% margin (like its y-axis range) is added to prevent from marker clipping at boundaries. To this end, I have introduced a function, independentAxisRangeMargin: this may only be used for Scatter plot but just made it as a util function for now.

Here are several screenshots I have captured: Before = no specific axis range; After = after setting axis range

- Histogram at Browse and Subset: Histogram filter plot (for comparison purpose with Histogram viz)
![histogramfilter-displayRange](https://user-images.githubusercontent.com/12802305/132718671-29bc654f-9c85-47ee-8303-7046ba4fe80a.png)

- Histogram visualization: now it is almost identical range with Histogram filter plot. Some differences in tick labels are probably due to the difference in plot size each other
![histogramViz-enrollmentDate](https://user-images.githubusercontent.com/12802305/132718733-e2c3cc6e-694a-45a9-aa42-01527ceb4215.png)

- Scatter plot number type: Before
![scatter-old-range1](https://user-images.githubusercontent.com/12802305/132718918-1f3522c3-f350-4add-bb93-f78d8608e46a.png)

- Scatter plot number type: After - some legend item was disabled by clicking it in order to show no changes at independent axis range
![scatter-new-range1](https://user-images.githubusercontent.com/12802305/132718955-3d278905-0a55-4837-a92a-0e642bdd587d.png)

- Scatter plot date type: Before
![scatter-date-old-range1](https://user-images.githubusercontent.com/12802305/132719167-4562fdac-9c72-4acd-acf1-fe1661163dec.png)

- Scatter plot date type: After - not that obvious but the upper bounds are slightly different each other: this shows slightly higher upper bound regardless of legend filtering
![scatter-date-new-range1](https://user-images.githubusercontent.com/12802305/132719176-835a5219-c0ef-4577-82b2-e230c586ffbd.png)

dmfalke commented on src/lib/core/components/visualizations/implementations/HistogramVisualization.tsx:
```diff
@@ -325,6 +330,20 @@ function HistogramViz(props: VisualizationProps) {
         independentAxisLabel={axisLabelWithUnit(xAxisVariable) ?? 'Main'}
+        // variable's metadata-based independent axis range with margin
+        independentAxisRange={
+          defaultIndependentRange && defaultIndependentRange != null
+            ? xAxisVariable?.type === 'date'
+              ? {
+                  min: defaultIndependentRange.min as string,
+                  max: defaultIndependentRange.max as string,
+                }
+              : {
+                  min: defaultIndependentRange.min as number,
+                  max: defaultIndependentRange.max as number,
+                }
+            : undefined
```
This should be computed in a `useMemo` hook to prevent excessive re-renders of the `Histogram` component.

moontrip replied:
@dmfalke useMemo() is now applied to both of them (this and below ones)

dmfalke commented on src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx:
```diff
@@ -371,6 +381,21 @@ function ScatterplotViz(props: VisualizationProps) {
           dependentAxisLabel={axisLabelWithUnit(yAxisVariable) ?? 'Y-Axis'}
+          // variable's metadata-based independent axis range with margin
+          independentAxisRange={
+            defaultIndependentRangeMargin &&
+            defaultIndependentRangeMargin != null
+              ? xAxisVariable?.type === 'date'
+                ? {
+                    min: defaultIndependentRangeMargin.min as string,
+                    max: defaultIndependentRangeMargin.max as string,
+                  }
+                : {
+                    min: defaultIndependentRangeMargin.min as number,
+                    max: defaultIndependentRangeMargin.max as number,
+                  }
+              : undefined
```
This should be computed in a `useMemo`, similar reason as above.

dmfalke reviewed (commented):
Please see my requested changes in two inline comments.

Also, if I'm not mistaken, the `axisMin` should be `min(dataRangeMin, annotatedRangeMin)`, and `axisMax` should be `max(dataRangeMax, annotatedDataRangeMax)`. The reason is that there might be cases where an annotated range is smaller than the data range itself, and we want to avoid hiding data by default. I don't see that logic being done here, but maybe I'm missing something?

moontrip commented:
> Please see my requested changes in two inline comments.
> 
> Also, if I'm not mistaken, the `axisMin` should be `min(dataRangeMin, annotatedRangeMin)`, and `axisMax` should be `max(dataRangeMax, annotatedDataRangeMax)`. The reason is that there might be cases where an annotated range is smaller than the data range itself, and we want to avoid hiding data by default. I don't see that logic being done here, but maybe I'm missing something?

@dmfalke thank you for your comments 👍 First of all, I will change those inline stuff to useMemo().

Speaking of min()/max() case, we did not use any specific data-based min/max for independent axis before: just relied on Plotly's autorange except histogram filter plot at `Browse and subset` tab. So, this implementation is purely based on variable's metadata min and max. Note that there may exist one or two kinds of range values at metadata such as displayRangeMin/Max and rangeMin/Max. Under current logic, displayRangeMin/Max has top priority and if not existing, then rangeMin/Max is used instead. Actually, this logic has been used at the histogram filter plot.

Histogram response offers two kinds of min/max, viewport min/max and data min/max: from my observation, yet limited, they correspond to metadata's min/max. However, scatter plot response does not offer such min/max statistics.

So, one question is, with an example of min, whether there exists a case that metadata's min is larger than data's min. Do you concern about this case ?

bobular commented:
What's the issue that this relates to?

moontrip commented:
> What's the issue that this relates to?

@bobular thank you for your interest. Perhaps you can answer Dave's question and check my answer above and offer some insights? 👍

bobular commented:
I will after you link it to an issue ;-)

dmfalke commented:
@moontrip can you link the issue to this PR? I think that will help with further discussions. Thanks!

moontrip commented:
@dmfalke @bobular Sorry about my confusion. IIRC, I did not miss to add `Linked Issues` before if exists, but somehow this was the case I missed it 😄

bobular commented:
> Speaking of min()/max() case, we did not use any specific data-based min/max for independent axis before: just relied on Plotly's autorange except histogram filter plot at `Browse and subset` tab. So, this implementation is purely based on variable's metadata min and max. Note that there may exist one or two kinds of range values at metadata such as displayRangeMin/Max and rangeMin/Max. Under current logic, displayRangeMin/Max has top priority and if not existing, then rangeMin/Max is used instead. Actually, this logic has been used at the histogram filter plot.
> 
> Histogram response offers two kinds of min/max, viewport min/max and data min/max: from my observation, yet limited, they correspond to metadata's min/max. However, scatter plot response does not offer such min/max statistics.
> 
> So, one question is, with an example of min, whether there exists a case that metadata's min is larger than data's min. Do you concern about this case ?

I think Dave means `rangeMin/Max` when he says `dataRangeMin/Max`

And your code seems to already do this.

On a call now - so will try to test this afterwards.

Is there a web-components branch with this?

moontrip commented:
> > Speaking of min()/max() case, we did not use any specific data-based min/max for independent axis before: just relied on Plotly's autorange except histogram filter plot at `Browse and subset` tab. So, this implementation is purely based on variable's metadata min and max. Note that there may exist one or two kinds of range values at metadata such as displayRangeMin/Max and rangeMin/Max. Under current logic, displayRangeMin/Max has top priority and if not existing, then rangeMin/Max is used instead. Actually, this logic has been used at the histogram filter plot.
> > Histogram response offers two kinds of min/max, viewport min/max and data min/max: from my observation, yet limited, they correspond to metadata's min/max. However, scatter plot response does not offer such min/max statistics.
> > So, one question is, with an example of min, whether there exists a case that metadata's min is larger than data's min. Do you concern about this case ?
> 
> I think Dave means `rangeMin/Max` when he says `dataRangeMin/Max`
> 
> And your code seems to already do this.
> 
> On a call now - so will try to test this afterwards.
> 
> Is there a web-components branch with this?

@bobular Thank you for your observation. This does not reply on web-components so you can test with this branch only 👍

dmfalke commented:
Sorry if I wasn't clear. What we want to avoid is any data being hidden (outside the viewport) by default. To accomplish this, for each axis, we would take have something like `axisMin = min(displayRangeMin, rangeMin)`, and `axisMax = max(displayRangeMax, rangeMax)`.

I don't see that being done anywhere in the code. It looks like `displayRangeMin` and `displayRangeMax` are always used. Again, I might be missing something.

moontrip commented:
> Sorry if I wasn't clear. What we want to avoid is any data being hidden (outside the viewport) by default. To accomplish this, for each axis, we would take have something like `axisMin = min(displayRangeMin, rangeMin)`, and `axisMax = max(displayRangeMax, rangeMax)`.
> 
> I don't see that being done anywhere in the code. It looks like `displayRangeMin` and `displayRangeMax` are always used. Again, I might be missing something.

@dmfalke Aha I see. Thank you for your clarification 👍 I thought that the range of displayRangeMin to Max is always larger than or equal to that of rangeMin to Max ? Histogram filter plot seems to follow this assumption(?) as they always prioritize displayRangeMin/Max over rangeMin/Max. Please correct me if I am wrong.

dmfalke commented:
In theory, it should be larger. But since displayRange{Min,Max} is entered by humans, it's possible this value can be smaller than the data's range. I am not aware of any such cases, but Danielle did mention it as a good practice to follow. Hope that makes sense!

moontrip commented:
> In theory, it should be larger. But since displayRange{Min,Max} is entered by humans, it's possible this value can be smaller than the data's range. I am not aware of any such cases, but Danielle did mention it as a good practice to follow. Hope that makes sense!

@dmfalke thank you for your detailed information 👍 I thought it was generated by a code or something like that, though it may still be incorrect. I agree that it makes much sense so I will add such a comparison as well.

bobular commented:
DK is right - the logic formerly used in HistogramFilter (L91 in `main` at the moment) is "use displayRangeMin/Max if available otherwise use rangeMin/Max".  That was probably my doing. 

The `axisMin = min(displayRangeMin, rangeMin)` approach certainly seems safer.

dmfalke commented:
Aha! I must have missed that detail. I'm wondering if there is an opportunity for both the filter and the viz components to share a common utility, to encapsulate this logic?

moontrip commented:
> Aha! I must have missed that detail. I'm wondering if there is an opportunity for both the filter and the viz components to share a common utility, to encapsulate this logic?

@dmfalke 👍 Will check this out as well if it would be feasible

moontrip commented:
@dmfalke (and @bobular) I have addressed suggestions: min/max comparison, changed util function, simplified codes, changed histogram filter plot's `independentAxisRange` with a util function used at vizs, etc.

dmfalke reviewed (approved):
Thanks @moontrip, this looks good to me! I didn't test the specifics of our most recent conversation, but the code looks good to me, and the general feature appears to be functioning as expected.

bobular reviewed (approved):
Looks good (see my one query) - tested using GEMS1A  (GEMS1 not available right now)

BUT there are no `displayRangeMin`-annotated variables in GEMS1A, so difficult to test.

I'll make a note in the original issue so that Outreach people test it properly.

moontrip commented on src/lib/core/components/filter/HistogramFilter.tsx:
```diff
@@ -78,6 +80,12 @@ export function HistogramFilter(props: Props) {
 
+  // compute default independent range from meta-data based util
+  const defaultIndependentRange: NumberOrDateRange | undefined = useMemo(
+    () => defaultIndependentAxisRange(variable),
```
Thank you for your review, @bobular 
The reason why the second argument like 'histogram' is due to the condition you might have set at histogram filter:

https://github.com/VEuPathDB/web-eda/blob/bbc6ea201b1c05d056a581db76953dcb6275c7da/src/lib/core/components/filter/HistogramFilter.tsx#L94

Honestly I don't know why it was set but though it meant something, so I made the second argument to respect the condition as I don't think it is necessary for scatter/line plot.

dmfalke replied:
That is based on the desire to use `0` as the origin for histograms (unless the range starts before `0`). This should probably have a comment.

I'm wondering if we should do the same for scatter plot (both axes). I can ask the question on slack.

bobular replied:
Oh, duh - I commented on an out-of-date commit :-(

dmfalke commented:
One more comment/question (sorry!): Should we apply this logic to _both_ axes of a scatter plot?

moontrip commented:
> One more comment/question (sorry!): Should we apply this logic to _both_ axes of a scatter plot?

@dmfalke If I remember correctly, there is no min/max for dependent axis (y-axis) from variable's metadata ?

dmfalke commented:
@moontrip for a scatter plot, each axis represents the range of values for each variable. So, one axis represents the range of values for the independent variable, and the other axis represents the range of values for the dependent variable. I think it makes sense to use the annotated range for both axes, not just the independent axis. This will provide further stability with scatter plots, as the subset changes. I suspect you can use the same utility function to compute the range for both axes. Does this make sense?

moontrip commented:
> @moontrip for a scatter plot, each axis represents the range of values for each variable. So, one axis represents the range of values for the independent variable, and the other axis represents the range of values for the dependent variable. I think it makes sense to use the annotated range for both axes, not just the independent axis. This will provide further stability with scatter plots, as the subset changes. I suspect you can use the same utility function to compute the range for both axes. Does this make sense?

@dmfalke Thank you for your detailed information! 👍 I got your point for dependent axis but at the same time it may cause additional trouble, especially dependent axis of scatter plot utilizes other sub-plots such as smoothed mean, confidence interval, and best fit lines. I think that those additional sub-plots are often out of variable's range, which may be a bit tricky to handle: and we have set 5% margin as well based on the max of all sub-plots.

dmfalke commented:
I see. Okay, let's see what others have to say. I suppose for now we should go ahead and merge this 🙂

moontrip commented:
> I see. Okay, let's see what others have to say. I suppose for now we should go ahead and merge this 🙂

@dmfalke Yeah we will see 😃 But either way, it is good to know that we can use Y-axis variable for min/max: didn't think of that case before.

moontrip commented:
I will merge this for now. Will handle additional works like X/Y range of scatter plot in another branch if they are demanded.
